src/squishy/core.clj

```diff
@@ -12,22 +14,46 @@
   (let [fail-message (pr-str {:body body :error (.getMessage error)})]
     (sqs/send client fail-queue-url fail-message)))
 
-(defn safe-process [client fail-queue-url f]
+(defn safe-process
+  [client queue-url fail-queue-url visibility-timeout f]
   (fn [message]
     (log/debug "Processing SQS message:" (str "<<" message ">>"))
-    (try (f message)
-      (catch Exception e
-        (let [body (:body message)]
-          (log/error "Failed to process" body e)
-          (report-error client fail-queue-url body e))))))
+    (let [processor (future
+                      (try (f message)
+                           (catch Exception e
+                             (let [body (:body message)]
+                               (log/error e "Failed to process" body)
+                               (report-error client fail-queue-url body e)))))]
+      (loop [timeout visibility-timeout]
+        (if (realized? processor)
+          @processor
+          (let [new-timeout (int (* timeout 1.5))]
+            (Thread/sleep (* 1000 (- timeout (/ timeout 4))))
```

Comment:
Hmm... won't this cause the processing of messages to slow down to a maximum rate of little better than 1 per `visibility-timeout`?

Say processing a message takes 1 second. It will not be ready the first time through; `(realized? processor)` will be false. Then this will pause for 22.5 seconds (in the default case) before checking again. Even though we could be processing more messages, we're blocking here for 22.5 seconds every time.

Perhaps this should be approached differently so as not to unnecessarily block processing.